File: fui_system/src/platform/qt/qt_wrapper/cpp/qsystemtrayicon.cpp
#include <QSystemTrayIcon>
#include "qsystemtrayicon.h"

void *QSystemTrayIcon_new()
{
    return static_cast<void *>(new (std::nothrow) QSystemTrayIcon());
}

void QSystemTrayIcon_delete(void *self)
{
    delete static_cast<QSystemTrayIcon *>(self);
}

void QSystemTrayIcon_setContextMenu(void *self, void *menu)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    QMenu *qmenu = static_cast<QMenu *>(menu);
    tray->setContextMenu(qmenu);
}

void QSystemTrayIcon_setIcon(void *self, const void *icon)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    const QIcon *qicon = static_cast<const QIcon *>(icon);
    tray->setIcon(*qicon);
}

void QSystemTrayIcon_setToolTip(void *self, const void *tip)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    const QString *qtip = static_cast<const QString *>(tip);
    tray->setToolTip(*qtip);
}

void QSystemTrayIcon_setVisible(void *self, int visible)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    tray->setVisible(visible != 0);
}

void QSystemTrayIcon_showMessage(void *self, const void *title, const void *message, int icon, int timeoutMilliseconds)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    const QString *qtitle = static_cast<const QString *>(title);
    const QString *qmessage = static_cast<const QString *>(message);
    tray->showMessage(*qtitle, *qmessage, (QSystemTrayIcon::MessageIcon)icon, timeoutMilliseconds);
}

void QSystemTrayIcon_showMessage2(void *self, const void *title, const void *message, const void *icon, int timeoutMilliseconds)
{
    QSystemTrayIcon *tray = static_cast<QSystemTrayIcon *>(self);
    const QString *qtitle = static_cast<const QString *>(title);
    const QString *qmessage = static_cast<const QString *>(message);
    const QIcon *qicon = static_cast<const QIcon *>(icon);
    tray->showMessage(*qtitle, *qmessage, *qicon, timeoutMilliseconds);
}

File: fui_system/src/platform/qt/qt_wrapper/qsystemtrayicon.rs
use crate::platform::qt::qt_wrapper::{QIcon, QMenu, QString};
use crate::FUISystemError;

pub struct QSystemTrayIcon {
    pub this: *mut ::std::os::raw::c_void,
}

impl QSystemTrayIcon {
    pub fn new() -> Result<Self, FUISystemError> {
        unsafe {
            let this = crate::platform::qt::qt_wrapper::QSystemTrayIcon_new();
            if this.is_null() {
                return Err(FUISystemError::OutOfMemory);
            }

            Ok(Self { this })
        }
    }

    // keeps reference to QMenu, doesn't take ownership
    pub fn set_context_menu(&mut self, menu: &mut QMenu) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_setContextMenu(self.this, menu.this);
        }
    }

    pub fn set_icon(&mut self, icon: &QIcon) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_setIcon(self.this, icon.this);
        }
    }

    pub fn set_tool_tip(&mut self, tip: &QString) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_setToolTip(self.this, tip.this);
        }
    }

    pub fn set_visible(&mut self, is_visible: bool) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_setVisible(
                self.this,
                if is_visible { 1 } else { 0 },
            );
        }
    }

    pub fn show_message(&mut self, title: &QString, message: &QString, icon: i32, timeout: i32) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_showMessage(
                self.this,
                title.this,
                message.this,
                icon,
                timeout,
            );
        }
    }

    pub fn show_message2(
        &mut self,
        title: &QString,
        message: &QString,
        icon: &QIcon,
        timeout: i32,
    ) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_showMessage2(
                self.this,
                title.this,
                message.this,
                icon.this,
                timeout,
            );
        }
    }
}

impl Drop for QSystemTrayIcon {
    fn drop(&mut self) {
        unsafe {
            crate::platform::qt::qt_wrapper::QSystemTrayIcon_delete(self.this);
        }
    }
}
